_plugins/git-metadata.rb

```diff
@@ -0,0 +1,165 @@
+require 'posix/spawn'
+
+module GitMetadata
+  def self.ctime(path)
+    return nil unless File.exist?(path)
+    File.ctime(path)
+  end
+  
+  def self.mtime(path)
+    return nil unless File.exist?(path)
+    File.mtime(path)
+  end
+  
+  def self.inject_dates(page)
+    # detect when we've got undated document injection working
+    raise "nil date for document" if page.data['date'].nil? and page.instance_of? Jekyll::Document
+
+    # Sometimes page.path is an absolute path, sometimes it's relative.
+    # Need it to always be relative for hash lookup.
+    relative_path = if page.path.start_with?('/')
+      Pathname.new(page.path)
+        .relative_path_from(
+          Pathname.new(File.dirname(Git.top_level_directory))
+        ).to_s
+    else
+      page.path
+    end
+
+    page.data['date'] = Git.files.dig(relative_path, :last_created_at) || ctime(relative_path) || Time.now if page.data['date'].nil?
+    page.data['modified_at'] = Git.files.dig(relative_path, :last_modified_at) || mtime(relative_path) || Time.now
+    page.data['commit'] = Git.files.dig(relative_path, :commit)
+  end
+
+  class Git
+    class << self
+
+      def files
+        @@files ||= populate_files
+      end
+
+      def branch
+        @@branch ||= Executor.sh(
+          "git",
+          "rev-parse",
+          "--abbrev-ref",
+          "HEAD"
+        ).strip
+      end
+
+      def init(site)
+        Jekyll.logger.debug "         Resetting: git cache"
+        @@site_source = site.source
+        @@branch = nil
+        @@files = nil
+      end
+
+      def populate_files
+        @@files = {}
+        timestamp = nil
+        commit = nil
+
+        Executor.sh(
+          'git',
+          '--git-dir',
+          top_level_directory,
+          'log',
+          '--name-only',
+          '--date=unix',
+          '--pretty=%%these-files-modified-at:%ct%n%%commit:%H'
+        ).split("\n").each do |line|
+          case
+          when line.empty? # skip
+          when line.start_with?('%these-files-modified-at:')
+            timestamp = line.split(':')[1]
+          when line.start_with?('%commit:')
+            commit =  line.split(':')[1]
+          else
+            @@files[line] = { last_modified_at: timestamp, commit: commit } unless @@files.key?(line)
+            @@files[line][:last_created_at] = timestamp
```

Comment:
Oh you know what this isn't right, this captures the "first created at" time—in other words it produces an unexpected value if a file is created, deleted, and then created again.

Should find a way to include file operations in the git log output and evict paths from `@@files` when they're deleted.

Reply:
This is fixed, but now it doesn't work properly for paths that contain spaces. I think my plan for now is to just not use spaces.